### PR TITLE
Move snapshot generation function to the snapshot tool

### PR DESCRIPTION
**Pull request by galpeter**
Now there is a snapshot tool which can merge snapshots
but was unable to generate snapshots by itself.

This change moves the snapshot generation utilities
to the snapshot tool.

JerryScript-DCO-1.0-Signed-off-by: Peter Gal pgal.u-szeged@partner.samsung.com

### CODE REVIEW

File: jerry-main/main-unix-snapshot.c [hunk 1]
```diff
@@ -73,6 +93,237 @@ read_file (uint8_t *input_pos_p, /**< next position in the input buffer */
...
+  if (snapshot_file_p == NULL)
+  {
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: Unable to write snapshot file: '%s'\n", output_file_name_p);
+    return JERRY_STANDALONE_EXIT_CODE_FAIL;
+  }
+
+  fwrite (output_buffer, sizeof (uint8_t), snapshot_size, snapshot_file_p);
+  fclose (snapshot_file_p);
+
+  printf ("Created snapshot file: '%s' (%d bytes)\n", output_file_name_p, snapshot_size);
```
Comment (akosthekiss) [inline jerry-main/main-unix-snapshot.c]: cppcheck signals signed/unsigned int mismatch here ("warning(invalidPrintfArgType_sint): %d in format string (no. 2) requires a signed integer given in the argument list.")

File: jerry-main/main-unix-snapshot.c [hunk 2]
```diff
@@ -73,6 +93,237 @@ read_file (uint8_t *input_pos_p, /**< next position in the input buffer */
...
+    if (literal_file_p == NULL)
+    {
+      jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: Unable to write literal file: '%s'\n", save_literals_file_name_p);
+      return JERRY_STANDALONE_EXIT_CODE_FAIL;
+    }
+
+    fwrite (output_buffer, sizeof (uint8_t), literal_buffer_size, literal_file_p);
+    fclose (literal_file_p);
+
+    printf ("Created literal file: '%s' (%d bytes)\n", save_literals_file_name_p, literal_buffer_size);
```
Comment (akosthekiss) [inline jerry-main/main-unix-snapshot.c]: ... and here

**Comment by zherczeg** [comment]
I like this change

**Comment by galpeter** [comment]
@akosthekiss I've updated the PR.

File: jerry-main/main-unix-snapshot.c [hunk 3]
```diff
@@ -73,6 +93,237 @@ read_file (uint8_t *input_pos_p, /**< next position in the input buffer */
...
+               .help = "print generated opcodes"),
+  CLI_OPT_DEF (.id = OPT_GENERATE_CONTEXT, .opt = "c", .longopt = "context",
+               .help = "specify the execution context of the snapshot: "
+                       "global or eval (default: global).", .meta = "MODE"),
+  CLI_OPT_DEF (.id = OPT_GENERATE_LITERAL_LIST, .longopt = "save-literals-list-format",
+               .meta = "FILE",
+               .help = "export literals found in parsed JS input (in list format)"),
+  CLI_OPT_DEF (.id = OPT_GENERATE_LITERAL_C, .longopt = "save-literals-c-format",
+               .meta = "FILE",
+               .help = "export literals found in parsed JS input (in C source format)"),
```
Comment (akosthekiss) [inline jerry-main/main-unix-snapshot.c]: the original `--save-snapshot-for-{global,eval}` options have already been rewritten to a new `--context={global,eval}` format. wouldn't it make sense to have something similar for `--save-literals-{c,list}-format`, too? e.g., as simple as `--save-literals={c,list}`.

File: tools/run-tests.py [hunk 4]
```diff
@@ -237,16 +237,22 @@ def run_jerry_debugger_tests(options):
 
     return ret_build | ret_test
 
+def update_test_args_with_snapshot_tool(jerry_path, test_args):
+    if "--snapshot-tool" in test_args:
+        snapshot_idx = test_args.index("--snapshot-tool")
+        test_args[snapshot_idx] = "--snapshot-tool=%s-snapshot" % (jerry_path)
```
Comment (akosthekiss) [inline tools/run-tests.py]: is this rewrite really needed? patching up the cli options this way is a bit creepy, and the same effect could be achieved by simply adding this "deduce snapshot tool path from engine path" logic to the run-test-suite.sh script. (explanation following there...)

File: tools/runners/run-test-suite.sh [hunk 5]
```diff
@@ -44,13 +44,21 @@ then
...
+if [[ "$1" =~ ^--snapshot-tool=.* ]]
 then
     TEST_FILES="$TEST_FILES.snapshot"
     TEST_FAILED="$TEST_FAILED.snapshot"
     TEST_PASSED="$TEST_PASSED.snapshot"
     IS_SNAPSHOT=true;
-    shift
+
+    SNAPSHOT_TOOL=${1#--snapshot-tool=}
+    SNAPSHOT_TOOL=(${SNAPSHOT_TOOL//,/ })
```
Comment (akosthekiss) [inline tools/runners/run-test-suite.sh]: ...following from above: you'd just have `SNAPSHOT_TOOL=$ENGINE-snapshot` here. (agreed, it looses a bit of flexibility by not allowing for snapshot tools placed anywhere in the file system for those who run this script manually. but who does get restricted by that, actually? run-test.py does force to have jerry and jerry-snapshot side-by-side anyway)

File: jerry-main/main-unix-snapshot.c (hunk 3 above)
Comment (galpeter) [reply jerry-main/main-unix-snapshot.c]: the logic is different, for --save-literals* you need to specify the output file as argument. This way you can generate the literal file and also the snapshot.

File: jerry-main/main-unix-snapshot.c [hunk 6]
```diff
@@ -73,6 +93,237 @@ read_file (uint8_t *input_pos_p, /**< next position in the input buffer */
...
+  (void) argc;
+
+  bool is_save_literals_mode_in_c_format = false;
+  bool is_snapshot_mode_for_global = true;
+  jerry_init_flag_t flags = JERRY_INIT_EMPTY;
+
+  uint32_t number_of_files = 0;
+  uint8_t *source_p = input_buffer;
+  size_t source_length = 0;
+  const char *output_file_name_p = "js.snapshot";
```
[the remark this answers is not in the document]
Comment (galpeter) [reply jerry-main/main-unix-snapshot.c]: IMHO having different default output names makes the intent more clear.
But we can change it to a common one. Currently merge uses "merged.snapshot",  so what do you think would be the best common name?

File: tools/runners/run-test-suite.sh (hunk 5 above)
Comment (galpeter) [reply tools/runners/run-test-suite.sh]: Certainly we can simply assume the path, but I wanted to give the test site runner more freedom if it is executed manually. 

But I'll update this part.

File: jerry-main/main-unix-snapshot.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (akosthekiss) [reply jerry-main/main-unix-snapshot.c]: "js.snapshot" is fine by me for both. it'd be a quite simple and elegant default.

File: jerry-main/main-unix-snapshot.c (hunk 3 above)
Comment (akosthekiss) [reply jerry-main/main-unix-snapshot.c]: oh that's right

File: tools/runners/run-test-suite.sh [hunk 7]
```diff
@@ -50,7 +50,14 @@ then
...
     IS_SNAPSHOT=true;
-    shift
+
+    SNAPSHOT_TOOL=${ENGINE}-snapshot
+    if [ ! -x $SNAPSHOT_TOOL ]
+    then
+        echo "$0: $SNAPSHOT_TOOL: not an executable"
+        exit 1
+    fi
+    shift;
```
Comment (akosthekiss) [inline tools/runners/run-test-suite.sh]: no need for the semicolon (shouldn't be used elsewhere either)

File: jerry-main/main-unix-snapshot.c [hunk 8]
```diff
@@ -74,6 +121,227 @@ read_file (uint8_t *input_pos_p, /**< next position in the input buffer */
...
+
+  fwrite (output_buffer, sizeof (uint8_t), snapshot_size, snapshot_file_p);
+  fclose (snapshot_file_p);
+
+  printf ("Created snapshot file: '%s' (%lu bytes)\n", output_file_name_p, (unsigned long) snapshot_size);
+
+  if (save_literals_file_name_p != NULL)
+  {
+    const size_t literal_buffer_size = jerry_parse_and_save_literals ((jerry_char_t *) source_p,
+                                                                       source_length,
```
Comment (akosthekiss) [inline jerry-main/main-unix-snapshot.c]: seems that the arg indentation is off by one space

File: tools/runners/run-test-suite.sh (hunk 7 above)
Comment (akosthekiss) [reply tools/runners/run-test-suite.sh]: semicolon still here but removed from after IS_SNAPSHOT=true?

Comment (akosthekiss) [reply tools/runners/run-test-suite.sh]: ping on this minor change request

**Pull request merged**